pages/mesosphere/dcos/1.14/administering-clusters/backup-and-restore/zookeeper/index.md

```diff
@@ -0,0 +1,93 @@
+---
+layout: layout.pug
+navigationTitle: ZooKeeper State
+title: Back up and Restore ZooKeeper
+menuWeight: 1
+excerpt: Backing up and restoring DC/OS ZooKeeper state
+render: mustache
+model: /mesosphere/dcos/1.14/data.yml
+--- 
+
+Back up the state of ZooKeeper inside a DC/OS cluster, and later restore from that backup.
+
+You may wish to back up the ZooKeeper state of your cluster before performing an upgrade or downgrade. You may need to restore your cluster to a known good state if something goes wrong during an upgrade. It is highly recommended to backup ZooKeeper state regularly to be prepared for the worst-case scenario.
+
+<p class="message--important"><strong>IMPORTANT: </strong>
+Restoring ZooKeeper from a backup should be the last resort to recover a DC/OS cluster. It is only applicable after confirming that the cluster has suffered permanent data loss including the ZooKeeper state.
+</p>
+
+# Explanation
+
+## Backup
+
+The ZooKeeper cluster within DC/OS is a system that provides distributed consensus between its nodes running on each master node. The ZooKeeper state can only progress once all nodes in the cluster have seen and agreed on a certain value. This implies that state of any one ZooKeeper node will contain the entire state information up until a certain point in time. Therefore backing up only one ZooKeeper node is sufficient to get reasonably close to the latest state for a ZooKeeper cluster backup. Creating the backup takes time and therefore in live system the backed up information will not be current. This problem could be avoided by stopping all ZooKeeper instances but that is not desirable in highly available system.
```

Comment:
> Creating the backup takes time and therefore in live system the backed up information will not be current. This problem could be avoided by stopping all ZooKeeper instances but that is not desirable in highly available system.

This sounds like a lose lose situation. How long does the backup take? What is the consequence of losing state transactions from the time the backup starts to when it finishes? Specifically, what information is stored in ZooKeeper related user workloads (which likely contain data the user cares about)?

Reply:
The backup may take up to several minutes, depending on how fast your disk is and how much you have stored in ZooKeeper. We had (malicious) ZooKeeper files in the 20GB range. The usual size should not exceed several hundred MB. It will still take a couple of seconds I guess.
The backup is intended for the worst-case scenario where a cluster is permanently broken.
The number one thing that users will care about in this scenario are the DC/OS service account private keys that are fairly static.
When you think creating a backup that is not current sounds bad but it is not. In fact it will be "more current" than you intended when you started the procedure in the first place.
Yes, it may be missing some more recent user workloads but anything that had been available at the time the backup was started will be captured.